### PR TITLE
Comment attachment : store on disk

### PR DESCRIPTION
**Pull request by Theodlz**
This PR aims to change the way attachments are stored. Right now, they are stored as bytes strings directly in the DB, in the `attachment_bytes` column of the `Comment` models.
In this PR, the model and the handler are modified, so that new attachments are stored to disk, but attachments that were added before these changes are still readable from the DB.

### What could be added?
We could add some code in the `GET` methods of the handlers that would move the attachments stored in the DB when a user `gets` a comment, to progressively migrate older comments from the DB to disk.
Otherwise, I am not sure how this is done, but maybe that could be done in the migration script? I'm not sure, and it seems that option one is maybe easier to implement and prevents a long migration.

Any suggestions are more than welcome!

### CODE REVIEW

File: skyportal/models/comment.py [hunk 1]
```diff
@@ -107,6 +127,76 @@ def construct_author_info_dict(self):
...
+        if self._attachment_path:
+            if os.path.exists(self._attachment_path):
+                # remove the file and other files in the same directory
+                os.remove(self._attachment_path)
+            parent_dir = Path(self._attachment_path).parent
+            try:
+                if parent_dir.is_dir():
+                    for file_name in os.listdir(parent_dir):
+                        file = str(parent_dir) + '/' + file_name
+                        print("delete_data file", file)
```
Comment (profjsb) [inline skyportal/models/comment.py:190]:
```suggestion
```

**Comment by stefanv** [comment]
Please note that we have no protocols in place to back up disk storage.

File: config.yaml.defaults [hunk 2]
```diff
@@ -605,3 +605,5 @@ analysis_services:
   analysis_folder: analysis_data
   sn_analysis_service:
     port: 6801
+
+comment_folder: comment_data
```
Comment (profjsb) [inline config.yaml.defaults:609]: This should be something like:
`persistentdata/comments`

Comment (profjsb) [reply config.yaml.defaults:609]: cf. https://github.com/skyportal/skyportal/pull/3505

**Review by profjsb — CHANGES_REQUESTED** [review]
Need to change the storage location to make use of the new persistent disk

File: config.yaml.defaults (hunk 2 above)
Comment (Theodlz) [reply config.yaml.defaults:609]: Hi Josh. Yes, I saw the PR! I'll change that soon.

**Comment by mcoughlin** [comment]
@guynir42 We should try to finish this one off.

**Review by guynir42 — APPROVED** [review]
Mostly minor comments. Looks like a really big improvement.

File: skyportal/models/comment.py [hunk 3]
```diff
@@ -110,6 +130,75 @@ def construct_author_info_dict(self):
...
+                # remove the file and other files in the same directory
+                os.remove(self._attachment_path)
+            parent_dir = Path(self._attachment_path).parent
+            try:
+                if parent_dir.is_dir():
+                    for file_name in os.listdir(parent_dir):
+                        file = str(parent_dir) + '/' + file_name
+                        if os.path.isfile(file):
+                            os.remove(file)
+                    parent_dir.rmdir()
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/models/comment.py:195]: So this directory contains any files associated with a given comment, so I think it makes sense.

File: skyportal/models/comment.py [hunk 4]
```diff
@@ -248,3 +337,12 @@ class CommentOnShift(Base, CommentMixin):
...
         doc="The Shift referred to by this comment.",
     )
+
+
+@event.listens_for(Comment, 'after_delete')
+@event.listens_for(CommentOnSpectrum, 'after_delete')
+@event.listens_for(CommentOnGCN, 'after_delete')
+@event.listens_for(CommentOnShift, 'after_delete')
+def delete_comment_data_from_disk(mapper, connection, target):
+    log(f'Deleting analysis data for analysis id={target.id}')
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/models/comment.py:347]: yes.

File: skyportal/handlers/api/comment_attachment.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+                application/json:
+                  schema: Error
+        """
+
+        with self.Session() as session:
+            stmt = sa.select(Comment).where(Comment.attachment_bytes.is_(None))
+            count_stmt = sa.select(func.count()).select_from(stmt.distinct())
+            total_missing = session.execute(count_stmt).scalar()
+
+            # get the number of Objs with CommentAttachment
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/handlers/api/comment_attachment.py:49]: oops.

File: skyportal/handlers/api/comment.py [hunk 6]
```diff
@@ -611,6 +610,9 @@ def post(self, associated_resource_type, resource_id):
 
             session.add(comment)
             session.commit()
+            if data_to_disk is not None:
+                comment.save_data(attachment_name, data_to_disk)
+                session.commit()
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/handlers/api/comment.py]: good idea.

**Pull request merged**